connector/src/main/java/org/jboss/as/connector/subsystems/resourceadapters/ResourceAdapterSubsystemParser.java

```diff
@@ -194,49 +194,81 @@ private void writeRaElement(XMLExtendedStreamWriter streamWriter, ModelNode ra,
         if (transactionSupport == TransactionSupportEnum.XATransaction) {
             isXa = true;
         }
-        if (ra.hasDefined(WM_SECURITY.getName()) && ra.get(WM_SECURITY.getName()).asBoolean()) {
-            streamWriter.writeStartElement(Activation.Tag.WORKMANAGER.getLocalName());
-            streamWriter.writeStartElement(WorkManager.Tag.SECURITY.getLocalName());
-            WM_SECURITY_MAPPING_REQUIRED.marshallAsElement(ra, streamWriter);
-            WM_SECURITY_DOMAIN.marshallAsElement(ra, streamWriter);
-            WM_ELYTRON_SECURITY_DOMAIN.marshallAsElement(ra, streamWriter);
-            WM_SECURITY_DEFAULT_PRINCIPAL.marshallAsElement(ra, streamWriter);
-            if (ra.hasDefined(WM_SECURITY_DEFAULT_GROUPS.getName())) {
-                streamWriter.writeStartElement(WM_SECURITY_DEFAULT_GROUPS.getXmlName());
-                for (ModelNode group : ra.get(WM_SECURITY_DEFAULT_GROUPS.getName()).asList()) {
-                    streamWriter.writeStartElement(WM_SECURITY_DEFAULT_GROUP.getXmlName());
-                    streamWriter.writeCharacters(group.asString());
-                    streamWriter.writeEndElement();
+        if (ra.hasDefined(WM_SECURITY.getName())) {
+            boolean securityEnabled;
+            boolean isExpression = false;
+
+            if (ra.get(WM_SECURITY.getName()).getType().equals(ModelType.EXPRESSION)) {
+                try {
+                    securityEnabled = ra.get(WM_SECURITY.getName()).resolve().asBoolean();
+                    isExpression = true;
+                } catch (Exception e) {
+                    securityEnabled = WM_SECURITY.getDefaultValue().asBoolean();
                 }
+            } else if (ra.get(WM_SECURITY.getName()).getType().equals(ModelType.BOOLEAN)) {
+                securityEnabled = ra.get(WM_SECURITY.getName()).asBoolean();
+            } else {
+                securityEnabled = WM_SECURITY.getDefaultValue().asBoolean();
+            }
+
+            // Preserves the expression in the XML if an expression was used to turn the WorkManager Security off
+            if (!securityEnabled && isExpression) {
+                streamWriter.writeStartElement(Activation.Tag.WORKMANAGER.getLocalName());
+                streamWriter.writeStartElement(WorkManager.Tag.SECURITY.getLocalName());
+                WM_SECURITY.marshallAsAttribute(ra, streamWriter);
+                streamWriter.writeEndElement();
```

Comment:
This is not correct.  The security element should be written if any of its attributes are defined, regardless of its value.

Reply:
The security element is written if the `wm-security` is either set to true or is an expression (regarding of its resolved value). Only in case the `wm-security` is false the element is not written at all. The enabled attribute is optional and is used only in case an expression was used to change the settings. The problem I see is if an expression was used the rest of the `wm-security` logic is not persisted and should be persisted in case the expression was used, regardless the actual value.